crates/types/src/rpc/transactions/l1_handler_transaction.rs

```diff
@@ -0,0 +1,128 @@
+use std::sync::Arc;
+
+use blockifier::transaction::transactions::L1HandlerTransaction as BlockifierL1HandlerTransaction;
+use starknet_api::core::{
+    ContractAddress as ApiContractAddress, EntryPointSelector as ApiEntryPointSelector,
+    Nonce as ApiNonce,
+};
+use starknet_api::transaction::{
+    Calldata as ApiCalldata, Fee as ApiFee, L1HandlerTransaction as ApiL1HandlerTransaction,
+    TransactionHash as ApiTransactionHash, TransactionVersion as ApiTransactionVersion,
+};
+use starknet_rs_core::crypto::compute_hash_on_elements;
+use starknet_rs_core::types::FieldElement;
+
+use crate::error::{DevnetResult, Error};
+use crate::felt::Felt;
+use crate::rpc::transactions::L1HandlerTransaction;
+use crate::traits::HashProducer;
+
+/// Cairo string for "l1_handler"
+const PREFIX_L1_HANDLER: FieldElement = FieldElement::from_mont([
+    1365666230910873368,
+    18446744073708665300,
+    18446744073709551615,
+    157895833347907735,
+]);
+
+impl L1HandlerTransaction {
+    /// Instanciates a new `L1HandlerTransaction`.
+    /// TODO: a bit too much arguments here...
+    pub fn with_hash(mut self, chain_id: Felt) -> Self {
+        self.transaction_hash = self.compute_hash(chain_id);
+        self
+    }
+
+    /// Computes the hash of a `L1HandlerTransaction`.
+    ///
+    /// # Arguments
+    ///
+    /// * `chain_id` - The chain ID.
+    pub fn compute_hash(&self, chain_id: Felt) -> Felt {
+        assert_eq!(
+            self.version,
+            FieldElement::ZERO.into(),
+            "L1 handler transaction only supports version 0"
+        );
+
+        // No fee on L2 for L1 handler transaction.
+        let fee = FieldElement::ZERO;
+
+        compute_hash_on_elements(&[
+            PREFIX_L1_HANDLER,
+            self.version.into(),
+            self.contract_address.into(),
+            self.entry_point_selector.into(),
+            compute_hash_on_elements(
+                &self
+                    .calldata
+                    .iter()
+                    .map(|felt| FieldElement::from(*felt))
+                    .collect::<Vec<FieldElement>>(),
+            ),
+            fee,
+            chain_id.into(),
+            self.nonce.into(),
+        ])
+        .into()
+    }
+
+    /// Creates a blockifier version of `L1HandlerTransaction`.
+    pub fn create_blockifier_transaction(&self) -> DevnetResult<BlockifierL1HandlerTransaction> {
+        let version: Felt = 0_u128.into();
+
+        let transaction = BlockifierL1HandlerTransaction {
+            tx: ApiL1HandlerTransaction {
+                contract_address: ApiContractAddress::try_from(self.contract_address)?,
+                entry_point_selector: ApiEntryPointSelector(self.entry_point_selector.into()),
+                calldata: ApiCalldata(Arc::new(self.calldata.iter().map(|f| f.into()).collect())),
+                nonce: ApiNonce(self.nonce.into()),
+                version: ApiTransactionVersion(version.into()),
```

Comment:
I've updated the code [here](https://github.com/glihm/starknet-devnet-rs/blob/7dd5d6e0fe8f14aca40ebd562cc3c5d5e16ce873/crates/types/src/rpc/transactions/l1_handler_transaction.rs#L41) and added a comment.

It's mostly because only version 0 is supported for now, and I wanted to ensure that this is limited to the value `0` until upgraded voluntarily by the developer.

If it's not good for you, I can definitely remove this assert. IMHO it's better to enforce the check as this is the only value that must be supported for now.

Reply:
So far we've avoided using `assert` in production code. We've relied on using `Err`, i.e. `Result`.

If we will have some kind of check, then why not use `self.version` but define a new variable in `create_blockifier_transaction`. That way we have to remember to apply the future change in two places (this method and `compute_hash`).

Also, why should the check be done only on hash computation (i.e. in `compute_hash`)?

Reply:
I'm ok with that, I'll change into an `Err` and that's a good catch, it perhaps make more sense to check it only inside the `create_blockifier_transaction` because we know that the blockifier doesn't support other that version 0.

As all the fields are public, having a constructor would be too much arguments and members can easily changed. I'm more in favor to add the version check in `create_blockifier_transaction`.

Sounds good for you?

Reply:
Fixed in bbedde4892b7ea48cccb8cd64c14dea47f906df3. At the end.. it's like other transaction types, the version is here, and it just need to be initialized correctly.
Perhaps the better way will be to add an enum if one day more version of L1HandlerTransaction are available.